### PR TITLE
main: get window size from gtk

### PR DESCRIPTION
**Pull request by ricardomv**
With this patch we can start wkline from wayland and the desktops widget will fail and return keeping the bar alive

### CODE REVIEW

**Comment by Lokaltog** [comment]
I agree we should depend more on GTK instead of XCB since it'll make implementing wayland support easier.

I'm not sure about removing the call to setting `_NET_WM_STRUT_PARTIAL`, this property tells the window manager to reserve space on the top of the screen, and it's necessary for some WMs (I had this issue with qtile). If we're going to support both wayland and X, wouldn't it be possible to throw in an `#if HAS_XCB` around the XCB calls?

**Comment by ricardomv** [comment]
if i would guess i would say gtk does the call to _NET_WM_STRUT_PARTIAL when you set GDK_WINDOW_TYPE_HINT_DOCK, can you check if you still have an issue in qtile with this patch

**Comment by Lokaltog** [comment]
IIRC the dock window hint doesn't actually dock the window anywhere if set as an EWMH property (I think it only ensures the window is present on every desktop or something like that). I didn't test setting it through GTK though, but I'll test this in qtile later and report back.

**Comment by ricardomv** [comment]
i installed qtile and it doesn't reserve the space but i think it's their problem because they handle the _WM_WINDOW_TYPE_HINT_DOCK they should also reserve the space

**Comment by Lokaltog** [comment]
According to [the spec](http://standards.freedesktop.org/wm-spec/latest/ar01s05.html):

> `_NET_WM_WINDOW_TYPE_DOCK` indicates a dock or panel feature. Typically a Window Manager would keep such windows on top of all other windows.

It doesn't specify that the WM should reserve space or place it at a particular location, and how would the WM know if the statusline is supposed to be located on the top or bottom of the desktop. The spec also says this about `_NET_WM_STRUT_PARTIAL`:

> The purpose of struts is to reserve space at the borders of the desktop. This is very useful for a docking area, a taskbar or a panel, for instance. The Window Manager should take this reserved area into account when coenstraining window positions - maximized windows, for example, should not cover that area.

Some window managers don't honor struts anyway (bspwm), but they provide a config setting for defining struts manually. qtile doesn't provide any way of reserving space other than honoring `_NET_WM_STRUT_PARTIAL`, so to maintain compatibility I think we should keep this setting.

I don't think it's possible to set this property with GTK (it's one of the reasons I ported the project to C from Python), but is it a problem to require XCB for X users and do it differently on wayland with a couple of `#ifdef` blocks?

**Comment by ricardomv** [comment]
ah you are right, im going to check if we can do it in gtk. i dont want #ifdef 's because they are hard to maintain and make the code almost unreadable

**Comment by ricardomv** [comment]
uff that was a bit harder than i expected

**Comment by Lokaltog** [comment]
Yeah, I've been struggling with this too. I think I had a similar solution before using xcb-wm, but I thought GDK used XCB or Xlib internally to set window properties?

**Comment by ricardomv** [comment]
yes it might but we don't care about those details and let gtk do it. That's why we use a toolkit instead of talking to X directly.

**Comment by ricardomv** [comment]
will it mess up things if i force a commit to my branch? i messed up some identation

**Comment by Lokaltog** [comment]
Allright!

As long as you use feature branches you won't mess anything up when rebasing or amending commits on that branch. If you're on the develop branch and rebase onto an old commit that's already in my develop branch your develop branch will detach and you'll screw up the commit log and I won't be able to merge the feature properly.

File: src/widgets/desktops.c [hunk 1]
```diff
@@ -162,8 +162,16 @@
...
+	xcb_connection_t *conn = xcb_connect(NULL, NULL);
+	if (xcb_connection_has_error(conn)) {
+		wklog("Could not connect to display %s.", getenv("DISPLAY"));
+		return 0;
+	}
+
+	/* FIXME ewmh should not be a global */
+	ewmh = malloc(sizeof(xcb_ewmh_connection_t));
+	xcb_intern_atom_cookie_t *ewmh_cookie = xcb_ewmh_init_atoms(conn, ewmh);
+	xcb_ewmh_init_atoms_replies(ewmh, ewmh_cookie, NULL);
```
Comment (Lokaltog) [inline src/widgets/desktops.c:174]: Don't you have to free `ewmh` and `conn`?

**Comment by ricardomv** [comment]
it's done, i think everything went well.
you should only free when you do a malloc, if someone allocated conn then it should free it too xcb_disconnect maybe? ~~and ewmh is a global witch i didn't allocated~~ ups yes i did, when we get it out of a global var we should free it
I just found [this](http://www.cs.uky.edu/~raphael/programming.html) about globals

**Comment by ricardomv** [comment]
@Lokaltog if it's ok for you this is ready to merge

**Comment by Lokaltog** [comment]
Sweet, excellent work! I'll review and merge it tomorrow morning.

**Pull request merged**